MismatchGeneration/1_mismatch_generations/Open Library Mismatches/OL-Publication-Date-Mismatch-Generation.py

```diff
@@ -0,0 +1,249 @@
+    else:
+        date_obj = datetime.strptime(date_str, "%B %Y")
+    # Convert the parsed date object to a string in ISO 8601 format.
+    iso_date_str = date_obj.strftime("%Y-%m-%dT%H:%M:%SZ")
+    return iso_date_str
+
+
+# In[7]:
+
+
+# Testing convert_to_datetime().
+date_str = "1996"
+print(f"1996 is {convert_to_datetime(date_str)} in datetime")
+date_str = "January 1996"
+print(f"January 1996 is {convert_to_datetime(date_str)} in datetime")
+date_str = "January 01, 1996"
+print(f"January 01, 1996 is {convert_to_datetime(date_str)} in datetime")
+date_str = "January 1, 1996"
+print(f"January 1, 1996 is {convert_to_datetime(date_str)} in datetime")
+
+
+# In[8]:
+
+
+mismatches_1 = []
+mismatches_2 = []
+mismatches_3 = []
+
+
+# In[9]:
+
+
+def find_mismatches(wikidata_data, dataframe):
+    """
+    This function accepts a dataframe with the item's Wikidata data and a dataframe which the
+    mismatches will be appended to, and crosschecks the item's publication date on Wikidata
+    with the data on the external source. If the corresponding item has no publication date on
+    the external source, it will not be appended to the dataframe as a mismatch.
+ 
+    Parameters:
+    wikidata_data: The pandas dataframe which contains the Wikidata ID, publication date,
+    and external source ID of the items of interest.
+    dataframe: The pandas dataframe which the mismatches will be appended to.
+    """
+    total_rows = len(wikidata_data)
+    for index, row in tqdm(wikidata_data.iterrows(), total=total_rows, desc="Processing"):
+        external_id = row["openLibraryID"]
+        wd_date = row["publicationDate"]
+        while True:
+            try:
+                pub_date = get_first_publication_date(external_id)
+                if pub_date is None:  # Skip if pub_date is None.
+                    print(f"Publication date not found on Open Library for item {row['item']}")
+                    break
+                try:
+                    ol_formatted_date = convert_to_datetime(pub_date)
+                except Exception as e:
+                    print(f"Error converting date for item {row['item']}: {e}")
+                    continue
+                if (wd_date != ol_formatted_date):
+                    try:
+                        date_str = datetime.strptime(wd_date, "%Y-%m-%dT%H:%M:%SZ").strftime("%Y-%m-%dT%H:%M:%SZ")
+                    except ValueError as e:
+                        print(f"Error: Time converting date for item {row['item']}: {e}")
+                        continue
+                    dataframe.append({
+                        "item_id": str(row["item"]),
+                        "statement_guid": str(""),
+                        "property_id": str("P577"),
+                        "wikidata_value": str(date_str),
+                        "meta_wikidata_value": str(""), 
+                        "external_value": str(pub_date),
+                        "external_url": str(""),
```

Comment:
Would definitely help if there was a link to the OpenLibrary page for the Wikidata item.  It's somewhat challenging to navigate through the 3 lists and trying to find the specific item from a specific list.